pysnn/learning.py

```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
+        defaults = {
+            "a_pre": a_pre,
+            "a_post": a_post,
+            "lr": lr,
+            "e_trace_decay": e_trace_decay,
+        }
+
+        super(MSTDPET, self).__init__(layers, defaults)
+
+    def update_state(self):
         r"""Update eligibility trace based on pre and postsynaptic spiking activity.
         
         This function has to be called manually after each timestep. Should not be called from within forward, 
         as this does is likely not called every timestep.
         """
 
-        self.e_trace *= self.e_trace_decay
-        self.e_trace += self.a_pre * torch.ger(
-            self.post_neuron.spikes.view(-1).float(), self.pre_neuron.trace.view(-1)
-        ) - self.a_post * torch.ger(
-            self.post_neuron.trace.view(-1), self.pre_neuron.spikes.view(-1).float()
-        )
+        for layer in self.layers:
+            # Update eligibility trace
+            layer["e_trace"] *= self.e_trace_decay
+            layer["e_trace"] += self.a_pre * torch.ger(
+                layer["post_spikes"].view(-1).float(), layer["pre_trace"].view(-1)
+            ) - self.a_post * torch.ger(
+                layer["post_trace"].view(-1), layer["pre_spikes"].view(-1).float()
+            )
+
+    def reset_state(self):
+        for layer in self.layers:
+            layer["e_trace"].fill_(0)
 
-    def forward(self, reward):
+    def step(self, reward):
         # TODO: add weight clamping?
-        self.connection.weight += self.lr * reward * self.e_trace
+        for layer in self.layers:
+            layer["weight"] += self.defaults["lr"] * reward * layer["e_trace"]
+
+
+#########################################################
+# Fede STDP
+#########################################################
+class FedeSTDP(LearningRule):
+    r"""STDP version for Paredes Valles, performs mean operation over the batch 
+    dimension before weight update."""
+
+    def __init__(self, layers, lr, w_init, a):
+        # Check layer formats
+        self.check_layers(layers)
+
+        # Set default hyper parameters
+        self.lr = lr
+        self.w_init = w_init
+        self.a = a
+
+        # To possibly later support groups, without chaning interface
+        defaults = {"lr": lr, "w_init": w_init, "a": a}
+
+        # Select only necessary paramters
+        for idx, layer in enumerate(layers):
+            new_layer = {}
+            new_layer["trace"] = layer["connection"]["trace"]
+            new_layer["weight"] = layer["connection"]["weight"]
+            layers[idx] = new_layer
+
+        super(FedeSTDP, self).__init__(layers, defaults)
+
+    def step(self):
+        for layer in self.layers:
+            w = layer["weight"]
+            trace = layer["trace"].view(-1, *w.shape)
+
+            # LTP and LTD
+            dw = w - self.w_init
+
+            # LTP computation
+            ltp_w = torch.exp(-dw)
+            ltp_t = torch.exp(trace) - self.a
```

Comment:
Trace should be normalized to [0, 1] by scaling with the current maximum at the moment of firing